graphql_query_derive/src/lib.rs

```diff
@@ -26,38 +27,48 @@ fn graphql_query_derive_inner(
 ) -> Result<proc_macro::TokenStream, syn::Error> {
     let input = TokenStream::from(input);
     let ast = syn::parse2(input)?;
-    let (query_path, schema_path) = build_query_and_schema_path(&ast)?;
-    let options = build_graphql_client_derive_options(&ast, query_path.clone())?;
-
-    generate_module_token_stream(query_path, &schema_path, options)
-        .map(Into::into)
-        .map_err(|err| {
-            syn::Error::new_spanned(
-                ast,
-                format!("Failed to generate GraphQLQuery impl: {}", err),
-            )
-        })
-}
 
-fn build_query_and_schema_path(input: &syn::DeriveInput) -> Result<(PathBuf, PathBuf), syn::Error> {
     let cargo_manifest_dir = env::var("CARGO_MANIFEST_DIR").map_err(|_err| {
         syn::Error::new_spanned(
-            input,
+            &ast,
             "Error checking that the CARGO_MANIFEST_DIR env variable is defined.",
         )
     })?;
 
-    let query_path = attributes::extract_attr(input, "query_path")?;
-    let query_path = format!("{}/{}", cargo_manifest_dir, query_path);
-    let query_path = Path::new(&query_path).to_path_buf();
-    let schema_path = attributes::extract_attr(input, "schema_path")?;
+    let schema_path = attributes::extract_attr(&ast, "schema_path")?;
     let schema_path = Path::new(&cargo_manifest_dir).join(schema_path);
-    Ok((query_path, schema_path))
+
+    let query_string = attributes::extract_attr(&ast, "query_string").ok();
+    let module_token_stream = match query_string {
+        Some(query_string) => {
+            let options = build_graphql_client_derive_options(&ast, None)?;
+            generate_module_token_stream_from_string(
+                query_string.as_str(),
+                schema_path.as_path(),
+                options,
+            )
+        }
+        None => {
+            let query_path = attributes::extract_attr(&ast, "query_path")?;
+            let query_path = format!("{}/{}", cargo_manifest_dir, query_path);
+            let query_path = Path::new(&query_path).to_path_buf();
+            let options = build_graphql_client_derive_options(&ast, Some(query_path.clone()))?;
+
+            generate_module_token_stream(query_path, &schema_path, options)
+        }
+    };
+
+    module_token_stream.map(Into::into).map_err(|err| {
+        syn::Error::new_spanned(
+            ast,
+            format!("Failed to generate GraphQLQuery impl: {}", err),
+        )
+    })
```

Comment:
Mega nit! But I’d argue that this is slightly more idiomatic
```suggestion
    Ok(module_token_stream.map_err(|err| {
        syn::Error::new_spanned(
            ast,
            format!("Failed to generate GraphQLQuery impl: {}", err),
        )?
    }).into()
```

Reply:
I'm here for the idiomatic nits since I don't write Rust too often. 😃 Thank you!

Reply:
Oh I think this is why the result is being mapped:

```
mismatched types
`proc_macro2::TokenStream` and `proc_macro::TokenStream` have similar names, but are actually distinct typesrustc[Click for full compiler diagnostic](rust-analyzer-diagnostics-view:/diagnostic%20message%20%5B0%5D?0#file%3A%2F%2F%2FUsers%2Fd%2Fsrc%2Fgithub.com%2Fgraphql-rust%2Fgraphql-client%2Fgraphql_query_derive%2Fsrc%2Flib.rs)
lib.rs(27, 6): expected `std::result::Result<proc_macro::TokenStream, syn::Error>` because of return type
lib.rs(185, 1): `proc_macro2::TokenStream` is defined in crate `proc_macro2`
lib.rs(79, 1): `proc_macro::TokenStream` is defined in crate `proc_macro`
lib.rs(61, 5): use `?` to coerce and return an appropriate `Err`, and wrap the resulting value in `Ok` so the expression remains of type `Result`: `Ok(`, `?)`
```

Reply:
Huh, I thought I tested this locally before commenting this. But again, feel free to ignore.